README.md
# BlinkReceipt Chrome Extension Library

This JavaScript library plugs in to a Chrome extension and enables the retrieval of order data from a set list of retailers.

## Setup

- Copy `accountLinking.js` into your extension's directory (CDN integration coming soon)

- You must configure your `manifest.json` in the following way (only relevant keys shown):

```
"background": {
        "scripts": [
            "accountLinking.js",
            "background.js"
        ]
    },
    "content_security_policy": "script-src 'self' 'wasm-eval'; object-src 'self'",
    "permissions": [
        "tabs",
        "storage",
        "cookies",
        "background",
        "*://*/*"
    ]
```
Note: These values are meant to _supplement_ any values you already have in your manifest for these keys.

Note: `accountLinking.js` must appear _before_ your background script(s) in the `background.scripts` key

## Interacting with the `AccountLinking` object

In your background script, you will interact with the `AccountLinking` object that is created when `accountLinking.js` is loaded. There are 2 functions you can use to retrieve orders:

1. `AccountLinking.getUserOrders(options, callback)`
    
2. `AccountLinking.enableScheduledGetOrders(interval, options, callback)`
    
The difference is that the former retrieves orders on-demand, whereas the latter sets up order fetching on a fixed interval.

For the scheduled call, the `interval` parameter is the duration of each retrieval interval in seconds.

The `options` parameter is an object with the following properties:

| Property | Type | Description |
|---|---|---|
| `clientKey`  | `string`  | Your license key  |
| `retailers` | `array<string>` | An array of the retailer names you would like to try to retrieve orders for. The list of supported retailers can be obtained by calling `AccountLinking.getSupportedRetailers()` |
| `cutOffDays` | `integer` | How far back (in days) in the user's order history to search |

The `callback` function has the following signature:

```
function callback(name, code, data, errorMessage)
```
and the parameters are as follows:

| Property | Type | Description |
|---|---|---|
| `name` | `string` | The name of the retailer for this set of retrieve orders |
| `code` | `integer` | The status code of this retrieval attempt (200 is OK) |
| `data` | `string` | A base-64 encoded, encrypted payload of the retrieval results, which should be decrypted server side using your secret key |
| `errorMessage` | `string` | If an error occurred, this will be populated with the error message |

## Decrypting Results

The `data` parameter will contain the hex-encoded encrypted results. Here is the procedure for decryption:
 - The algorithm used for encryption & decryption is AES-128-CBC
 - The key is the first 16 bytes of your _secret key_ as assigned during setup
 - The initialization vector is the first 16 bytes of the encrypted text
 - You should remove the IV from the ciphertext prior to decryption (i.e. pull out the first 16 bytes which are the IV, and use the rest as the input for decryption)
 
### Decryption Samples in Different Languages 
- Node.js
```javascript
function decryptData(text, secretKey) {
    //initialize ciphertext from hex-encoded input
    const buffer = Buffer.from(text, 'hex');

    //construct decryption key as first 16 bytes of secretKey
    const finalKey = Buffer.from(secretKey).slice(0, 16).toString();
    
    //set iv to first 16 bytes of ciphertext
    const iv = new Buffer.from(buffer.slice(0, 16));

    //remove first 16 bytes from ciphertext
    const encryptedText = new Buffer.from(buffer.slice(16, buffer.length));

    //perform decryption
    const decipher = crypto.createDecipheriv("aes-128-cbc", finalKey, iv);
    let decrypted = decipher.update(encryptedText);
    decrypted = Buffer.concat([decrypted, decipher.final()]);
    return decrypted.toString();
}
```
- Java
```java
import java.security.Key;
import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.util.Arrays;
import org.apache.commons.codec.binary.Hex;
public class Decrypt {
    public static void main(String[] args) {
        try {
            String hex = "encrypted_data"
            byte[] bytes = Hex.decodeHex(hex.toCharArray());
            byte[] iv = Arrays.copyOfRange(bytes, 0, 16);
            byte[] text = Arrays.copyOfRange(bytes, 16, bytes.length);
            
            String key = "my_secret_key";
            byte[] finalKey = Arrays.copyOfRange(key.getBytes(), 0, 16);
            Key aesKey = new SecretKeySpec(finalKey, "AES");
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            IvParameterSpec ivParameterSpec = new IvParameterSpec(iv);
            cipher.init(Cipher.DECRYPT_MODE, aesKey, ivParameterSpec);
            String decrypted = new String(cipher.doFinal(text));
            System.err.println("Decrypted: " + decrypted);
        } catch(Exception e) {
          e.printStackTrace();
        }
    }
}
```
- Shell script
```bash
#!/bin/bash
text="encrypted_text_here"
mySecretKey="my_secret_key_here"
iv=${text:0:32}
payload=${text:32}
key=$(xxd -pu <<< "${mySecretKey:0:16}")
echo `echo $payload | xxd -r -p | openssl enc -aes-128-cbc -d -v -p -nosalt -K $key -iv $iv`
```

## Results Structure

Once decrypted, the results will be a JSON-encoded string that deserializes into this structure:

### Top Level object

| Key | Type | Notes |
|---|---|---|
| `merchantName` | `string` | |
| `total` | `float` | |
| `subtotal` | `float` | |
| `taxes` | `float` | |
| `receiptDate` | `string` | Format is `mm/dd/yyyy` |
| `receiptTime` | `string` | Format is `hh:mm` |
| `transactionId` | `string` | |
| `storeNumber` | `string` | |
| `storeAddress` | `string` | |
| `barcode` | `string` | |
| `last4CC` | `string` | |
| `products` | `array<Product>` | |
| `shipments` | `array<Shipment>` | `Walmart` & `Amazon` only|
| `paymentMethods` | `array<PaymentMethod>` | |
| `qualifiedPromos` | `array<Promotion>` | |

<br /><br />

### `Product` object

| Key | Type | Notes |
|---|---|---|
| `productNumber` | `string` | |
| `productName` | `string` | |
| `brand` | `string` | |
| `quantity` | `float` | |
| `unitOfMeasure` | `string` | e.g. "ounces", "packages", etc |
| `size` | `string` | |
| `unitPrice` | `float` | |
| `totalPrice` | `float` | The total price paid for this item taking into account quantities and discounts |
| `fullPrice` | `float` | The full price of this item before any discounts or coupons were applied |
| `category` | `string` | |
| `imgUrl` | `string` | |
| `upc` | `string` | |
| `shippingStatus` | `string` | |

<br /><br />

### `PaymentMethod` object

| Key | Type | Notes |
|---|---|---|
| `amount` | `float` | |
| `method` | `string` | e.g. "Cash", "Credit", etc |
| `cardType` | `string` | e.g. "Visa", "Mastercard", etc |

<br /><br />

### `Shipment` object

| Key | Type | Notes |
|---|---|---|
| `fulfillmentType` | `string` | `delivery`, `curbside pickup`, etc... |
| `status` | `string` | `canceled`, `delivered`, etc... |
| `products` | `array<Product>` |  |

<br /><br />

### `Promotion` object

| Key | Type | Notes |
|---|---|---|
| `slug` | `string` | The identifier for this promo |
| `rewardValue` | `float` | |
| `rewardCurrency` | `string` | e.g. "USD" |
| `relatedProductIndexes` | `array<integer>` | The indexes of the items in the `products` array that qualified for this promotion |
| `qualifications` | `array<array<integer>>` | An array containing one element for each instance where this promotion qualified in an order. For example, a promo to buy ketchup & mustard would qualify twice on an order with `products` "ketchup, mustard, ketchup, mustard" and this property would have value: `[ [0,1], [2,3] ]` |

CHANGELOG.md
# Change Log

## 1.2.5

#### Fixed
* Some time we return 200 while the parsing actually failed.

---
## 1.2.4

#### Added
* Add coupons support

#### Fixed
* ProdIntel improvements

---
## 1.2.3

#### Fixed
* Moving amazon order requests to the background script

---
## 1.2.2

#### Added
* Returning partial orders with `orderUrl` value for `Amazon` when extra login is required to open the order details
* ProdIntel improvements
* Bug fixes and optimizations

---
## 1.2.1

#### Added
* `Picked Up` shipment status

---
## 1.2.0

#### Added
* `getMerchantStatus` function to check if user's session is valid or not
* shipment_status returns one of the following options: `Ordered`, `Delivered`, `Undelivered`, `Shipped`, `Returned`, `Cancelled`, `Invalid`

---
## 1.1.0

#### Added
* Error code improvements. Return 204 when list is empty.

#### Fixed
* `Amazon`
    * WholeFoods order support 
* `Shipt`
    * Unable to authenticate user 
* `Meijer`
    * Parsing orders 
    

---
## 1.0.1

#### Added
* `Walmart`
    * Split products into shipments 
* `Amazon`
    * Split products into shipments 
    * Added order subtotal and taxes 

#### Fixed
* `Walmart`
    * Adding support for the new walmart orders list


---
## 1.0.0
- Initial release


demo/popup.js
document.addEventListener('DOMContentLoaded', () => {
    loadOrders();

    function loadOrders() {

        // chrome.runtime.getBackgroundPage(bg => {
                var bg = chrome.extension.getBackgroundPage();
                var ordersMap = bg._ordersCache || {};
                var ordersEl = document.getElementById("orders-list");
                let ordersListStr = "";

                for (let retailerId in ordersMap) {
                    let retailerMap = ordersMap[retailerId];
                    let isFirstOrder = false;

                    for (let orderKey in retailerMap) {
                        
                        let order = retailerMap[orderKey];
                        let uniqueValue = order[orderKey];
                       
                        let orderProductsCount = 0;
                        const products = order.products || null;
                        const shipments = order.shipments || null;
                        if (products != null) {
                            orderProductsCount += products.length;
                        } else if (shipments != null) {
                            for (let i = 0; i < shipments.length; i++) {
                                const shipment = shipments[i] || {};
                                const products = shipment.products || [];
                                orderProductsCount += products.length;
                            }
                        }
            
                        let date = new Date(order.receiptDate || order.dateUTC);
                        const ye = new Intl.DateTimeFormat('en', { year: 'numeric' }).format(date);
                        const mo = new Intl.DateTimeFormat('en', { month: 'short' }).format(date);
                        const da = new Intl.DateTimeFormat('en', { day: '2-digit' }).format(date);
                        console.log(`${da}-${mo}-${ye}`);
    
                        let merchantName = order.merchantName || "";
                        if (retailerId == "8897") {
                            merchantName = "Walmart Grocery";
                        }

                        if (isFirstOrder === false) {
                            isFirstOrder = true;
                            ordersListStr +=
                            `<li class="order-header">
                                <span style="text-align:left;">${merchantName}</span>
                            </li>`;
                        }

                        ordersListStr += 
                        `
                        <li class="amazon-order">
                            <div>
                                <span><strong>Order #:</strong> ${orderKey}</span>
                                <span>$${order.total}</span>
                            </div>
                            <div>
                                <span><strong>Date:</strong> ${mo} ${da}, ${ye}</span>
                                <span>${orderProductsCount} products</span>
                            </div>
                        </li>`
                    }
                }
                ordersEl.innerHTML = ordersListStr;
            // });

        setTimeout(function() {
            loadOrders()
        }, 1000);
    }

    // var bg = chrome.extension.getBackgroundPage();
    // var orders = bg.foundAmazonOrders || [];
    
    // var ordersEl = document.getElementById("amazon-orders-list");
    // for (let orderNum in orders) {

    //     const order = orders[orderNum];

    //     let li = document.createElement('li');
    //     li.className = "amazon-order";
    //     li.textContent = orderNum;
    //     ordersEl.appendChild(li);   

    // }

});


demo/options.js
document.addEventListener('DOMContentLoaded', () => {
    let el = document.getElementById("global-cutoff-days");
    
    chrome.storage.sync.get(settings => {
        document.getElementById("global-cutoff-days").value = settings['alp-global-cutoffdays'] || 30;
        const fetchMode = settings['run-mode'] || 'on-tab-open';
        resetRadioBut(fetchMode);
    });

    el.addEventListener('mouseup', event => {
        // console.log(event.target.value);
        chrome.storage.sync.set({'alp-global-cutoffdays': event.target.value});
    });
    el.addEventListener('keyup', event => {
        // console.log(event.target.value || 0);
        chrome.storage.sync.set({'alp-global-cutoffdays': event.target.value});
    });

    let fetchSecEl = document.getElementById("scheduled-secs");
    fetchSecEl.addEventListener('mouseup', event => {
        // console.log(event.target.value);
        chrome.storage.sync.set({'scheduled-secs': event.target.value});
    });
    fetchSecEl.addEventListener('keyup', event => {
        // console.log(event.target.value || 0);
        chrome.storage.sync.set({'scheduled-secs': event.target.value});
    });

    // ------------------------------------
    let radioElements = document.getElementsByClassName('radio');
    for (let i = 0; i < radioElements.length; i++) {
        let radioEl = radioElements[i];
        radioEl.addEventListener('change', event => {
            console.log(event.target.value || 'on-tab-open');
            resetRadioBut(event.target.id);
            chrome.storage.sync.set({'run-mode': event.target.value || 'on-tab-open'});
        });
    }

    function resetRadioBut(id) {
        let radioElements = document.getElementsByClassName('radio');
        for (let i = 0; i < radioElements.length; i++) {
            let radioEl = radioElements[i];
            if (radioEl.id == id) {
                radioEl.checked = true
                continue;
            }
            radioEl.checked = false
        }
    }

});


demo/background.js
const _fetchModes = {
    ON_TAB_OPEN: 'on-tab-open',
    SCHEDULED: 'scheduled'
}
var _ordersCache = {};
let _fetchMode = "on-tab-open";
let _supportedRetailers = null;

chrome.storage.sync.get(settings => {
    _fetchMode = settings['run-mode'] || _fetchModes.ON_TAB_OPEN;

    // Setup observers
    chrome.storage.onChanged.addListener((changes, namespace) => {
        for (key in changes) {
            let storageChange = changes[key];
            let newValue = storageChange.newValue;
            if (key == 'run-mode') {
                _fetchMode = newValue;
                if (newValue == _fetchModes.SCHEDULED) {
                    chrome.storage.sync.get(settings => {
                        const scheduledSecs = parseInt(settings['scheduled-secs']) || 10;
                        let params = { retailers: validRetailers(null) };
                        getDefaultOpts(opts => {
                            AccountLinking.enableScheduledGetOrders(scheduledSecs, {...params, ...opts}, processOrders);
                        });
                    });
                    return;
                }
                AccountLinking.disableScheduledGetOrders();
            } else if (key == 'scheduled-secs') {
                if (_fetchMode == _fetchModes.SCHEDULED) {
                    let params = { retailers: validRetailers(null) };
                    getDefaultOpts(opts => {
                        AccountLinking.enableScheduledGetOrders(scheduledSecs, {...params, ...opts}, processOrders);
                    });
                }
            }
        }
    });
    
    chrome.tabs.onActivated.addListener(tabInfo => {
        if (_fetchMode == _fetchModes.ON_TAB_OPEN) {
            chrome.tabs.get(tabInfo.tabId, (tab) => {
                getUserOrders({
                    url: tab.url,
                    fetchMode: _fetchMode,
                    retailers: validRetailers(tab.url)
                });
            });
        }
    });
    
    chrome.tabs.onUpdated.addListener(function(tabId, changeInfo, tab) {
        // if (_fetchMode == _fetchModes.ON_TAB_OPEN) {
        //     getUserOrders({
        //         url: tab.url,
        //         fetchMode: _fetchMode,
        //         retailers: validRetailers(tab.url)
        //     });
        // }
    });

    if (_fetchMode == _fetchModes.SCHEDULED) {
        const scheduledSecs = parseInt(settings['scheduled-secs']) || 10;
        let params = { retailers: validRetailers(null) };
        getDefaultOpts(opts => {
            AccountLinking.enableScheduledGetOrders(scheduledSecs, {...params, ...opts}, processOrders);
        }); 
    }

});

function validRetailers(url) {
    if (_supportedRetailers == null) {
        _supportedRetailers = AccountLinking.getSupportedRetailers();
    }
    if (_fetchMode == _fetchModes.SCHEDULED) {
        return _supportedRetailers;
    }
    const finalList = _supportedRetailers.filter(name => {
        if (name == 'Amazon') {
            // return (url || '').includes(`${name.toLowerCase()}.`) && (url || '').includes('order-history');   
        }
        return (url || '').includes(`${name.toLowerCase()}.`);
    });
    console.log(finalList);
    return finalList;
}

function getDefaultOpts(callback) {
    let opts = {
        clientKey: 'YOUR-KEY-HERE'
    }
    chrome.storage.sync.get(settings => {
        opts.cutOffDays = parseInt(settings['alp-global-cutoffdays']) || 60;
        callback(opts);
    });
}

function getUserOrders(params) {
    if ((params.retailers || []).length == 0) {
        return;
    }

    let validMerchants = params.retailers || [];
    let counter = validMerchants.length;
    for (const validMerchant of validMerchants) {
        AccountLinking.getMerchantStatus(validMerchant, (status) => {
            console.log(`${validMerchant} status: ${status} `);
            counter -= 1;
            if (counter <= 0) {
                params.ignoreOrdersCache = true;
                getDefaultOpts(opts => {
                    AccountLinking.getOrders({...params, ...opts}, processOrders);
                });
            }
        });
    }   
    
}

function processOrders(name, code, data, errorMessage) {
    console.log('----------------------------');
    console.log(`Code ${code}, Error: ${errorMessage}`);
    if (code != 200 && code != 204) {
        return;
    }
    
    console.log(data);
    const orders = JSON.parse(data);
    if (orders == null) {
        console.log(data);
    } else {
        console.log('Orders: ');
        console.log(orders);
    }
    
    for (let x = 0; x < orders.length; x++) {
        const order = orders[x];
        const retailerId = order.retailerId;
        const uniqueValue = order['udid'];
        let retailerOrdersMap = _ordersCache[retailerId] || {};
        retailerOrdersMap[uniqueValue] = order;
        _ordersCache[retailerId] = retailerOrdersMap; 
    }
    console.log(_ordersCache);
}